src/psyclone/tests/kernel_transformation_tests.py

```diff
@@ -0,0 +1,100 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+
+''' Module containing tests for kernel transformations. '''
+
+from __future__ import absolute_import
+import os
+import pytest
+from utils import get_invoke
+from psyclone.transformations import TransformationError
+
+
+def test_accroutine_err():
+    ''' Check that we raise the expected error if we can't find the
+    source of the kernel subroutine. '''
+    from psyclone.psyGen import Kern
+    from psyclone.transformations import ACCRoutineTrans
+    import fparser
+    _, invoke = get_invoke(os.path.join("dynamo0p3", "1_single_invoke.f90"),
+                           api="dynamo0.3", idx=0)
+    sched = invoke.schedule
+    kernels = sched.walk(sched.children, Kern)
+    kern = kernels[0]
+    assert isinstance(kern, Kern)
+    # Edit the fparser1 AST of the kernel so that it does not have a
+    # subroutine of the correct name
+    ast = kern._module_code
+    mod = ast.content[0]
+    # Find the subroutine statement
+    for child in mod.content:
+        if isinstance(child, fparser.one.block_statements.Subroutine):
+            sub = child
+    # Find the end subroutine statement
+    for child in sub.content:
+        if isinstance(child, fparser.one.block_statements.EndSubroutine):
+            end = child
+    sub.name = "some_other_name"
+    end.name = sub.name
+    rtrans = ACCRoutineTrans()
+    with pytest.raises(TransformationError) as err:
+        _ = rtrans.apply(kern)
+    assert ("Failed to find subroutine source for kernel testkern_code"
+            in str(err))
+
+
+def test_accroutine():
+    ''' Test that we can transform a kernel by adding a "!$acc routine"
+    directive to it. '''
+    from psyclone.gocean1p0 import GOKern
+    from psyclone.transformations import ACCRoutineTrans
+    from fparser.two import Fortran2003
+    _, invoke = get_invoke(os.path.join("gocean1p0", "nemolite2d_alg_mod.f90"),
+                           api="gocean1.0", idx=0)
+    sched = invoke.schedule
+    kern = sched.children[0].children[0].children[0]
+    assert isinstance(kern, GOKern)
+    rtrans = ACCRoutineTrans()
+    assert rtrans.name == "ACCRoutineTrans"
+    new_kern, _ = rtrans.apply(kern)
+    # The transformation should have populated the fparser2 AST of
+    # the kernel...
+    assert new_kern._fp2_ast
+    assert isinstance(new_kern._fp2_ast, Fortran2003.Program)
+    # Check AST contains directive
+    comments = Fortran2003.walk_ast(new_kern._fp2_ast.content,
+                                    [Fortran2003.Comment])
+    assert len(comments) == 1
+    assert str(comments[0]) == "!$acc routine"
```

Comment:
Test that it has inserted it in the expected (correct) place.

Reply:
Good point. Done.